be/src/vec/exprs/vcompound_pred.h

```diff
@@ -53,6 +53,75 @@ class VCompoundPred : public VectorizedFnCall {
 
     const std::string& expr_name() const override { return _expr_name; }
 
+    bool is_all_ones(const roaring::Roaring& r) {
+        return r.contains(0);
+        for (roaring::RoaringSetBitForwardIterator i = r.begin(); i != r.end(); ++i) {
+            if (*i == 0) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    //   1. when meet 'or' conjunct: a or b, if b can apply index, return all rows, so b should not be extracted
+    //   2. when meet 'and' conjunct, function with column b can not apply inverted index
+    //      eg. a and hash(b)=1, if b can apply index, but hash(b)=1 is not for index, so b should not be extracted
+    //          but a and array_contains(b, 1), b can be applied inverted index, which b can be extracted
+    Status eval_inverted_index(
+            VExprContext* context,
+            const std::unordered_map<ColumnId, std::pair<vectorized::NameAndTypePair,
+                                                         segment_v2::InvertedIndexIterator*>>&
+                    colId_to_inverted_index_iter,
+            uint32_t num_rows, roaring::Roaring* bitmap) const override {
+        if (_op == TExprOpcode::COMPOUND_OR) {
+            for (auto child : _children) {
+                std::shared_ptr<roaring::Roaring> child_roaring =
+                        std::make_shared<roaring::Roaring>();
+                Status st = child->eval_inverted_index(context, colId_to_inverted_index_iter,
+                                                       num_rows, child_roaring.get());
+                if (!st.ok()) {
+                    continue;
+                }
+                *bitmap |= *child_roaring;
+                if (!child_roaring->isEmpty()) {
```

Comment:
I think the check for NOT empty is wrong. And further, short circuit for OR is different to AND, you should check there is no zero in the total bitmap and skip remaining exprs.

Reply:
use  cardinality == rows to make sure if there are all rows here , no need to eval next expr